examples/functorch/parallel_train_torchopt.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2022 MetaOPT Team. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import argparse
+import math
+from collections import namedtuple
+from typing import Any, NamedTuple
+
+import functorch
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from functorch import combine_state_for_ensemble, grad_and_value, make_functional, vmap
+
+import torchopt
+
+
+def make_spirals(n_samples, noise_std=0.0, rotations=1.0):
+    ts = torch.linspace(0, 1, n_samples, device=DEVICE)
+    rs = ts**0.5
+    thetas = rs * rotations * 2 * math.pi
+    signs = torch.randint(0, 2, (n_samples,), device=DEVICE) * 2 - 1
+    labels = (signs > 0).to(torch.long).to(DEVICE)
+
+    xs = rs * signs * torch.cos(thetas) + torch.randn(n_samples, device=DEVICE) * noise_std
+    ys = rs * signs * torch.sin(thetas) + torch.randn(n_samples, device=DEVICE) * noise_std
+    points = torch.stack([xs, ys], dim=1)
+    return points, labels
+
+
+class MLPClassifier(nn.Module):
+    def __init__(self, hidden_dim=32, n_classes=2):
+        super().__init__()
+        self.hidden_dim = hidden_dim
+        self.n_classes = n_classes
+
+        self.fc1 = nn.Linear(2, self.hidden_dim)
+        self.fc2 = nn.Linear(self.hidden_dim, self.n_classes)
+
+    def forward(self, x):
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.fc2(x)
+        x = F.log_softmax(x, -1)
+        return x
+
+
+class ParallelTrainFunctorchOriginal:
+    def __init__(self, loss_fn, lr):
+        self.loss_fn = loss_fn
+        self.lr = lr
+        self.func_model, _ = make_functional(MLPClassifier().to(DEVICE))
+
+    def init_fn(self, num_models):
+        models = [MLPClassifier().to(DEVICE) for _ in range(num_models)]
+        _, batched_weights, _ = combine_state_for_ensemble(models)
+        return batched_weights
+
+    def train_step_fn(self, weights, batch, targets):
+        def compute_loss(weights, batch, targets):
```

Comment:
Can't we make a separate method with this, rather than defining a fun in a method?

Reply:
Same here, just mimicking the way zou did in functorch [parallel training examples](https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L78-L93). Also, similar way to make it functional can be found in functorch [MAML omniglot example](https://github.com/pytorch/functorch/blob/main/examples/maml_omniglot/maml-omniglot-transforms.py#L125-L147).

Reply:
I think in general Jax also, just [one example](https://huggingface.co/spaces/dalle-mini/dalle-mini/blob/53dade7fd869d7f66821a7179e9f508655b9afd0/tools/train/train.py#L553-L580) for demonstration. May be that's jax style :)

Reply:
But I think loss_fn can be indeed defined outside train_step_fn.

Reply:
Either way works, we didn't think too much about code clarity when we wrote https://github.com/pytorch/functorch/blob/main/examples/ensembling/parallel_train.py#L78-L93